crates/rustc_codegen_spirv/src/codegen_cx/entry.rs

```diff
@@ -121,14 +139,41 @@ impl<'tcx> CodegenCx<'tcx> {
         // Adjust any global `OpVariable`s as needed (e.g. loading from `Input`s).
         let arguments: Vec<_> = declared_params
             .iter()
-            .zip(&entry_fn_abi.args)
+            .zip(arg_abis)
             .zip(hir_params)
-            .map(|((&(var, storage_class), entry_fn_arg), hir_param)| {
+            .flat_map(|((&(var, storage_class), entry_fn_arg), hir_param)| {
                 match entry_fn_arg.layout.ty.kind() {
-                    TyKind::Ref(..) => var,
-
+                    TyKind::Ref(_, ty, _) if ty.is_sized(self.tcx.at(span), self.param_env()) => iter::once(var).chain(None),
+                    TyKind::Ref(_, ty, _) => {
+                        match ty.kind() {
+                            TyKind::Adt(adt_def, substs) => {
+                                let (member_idx, field_def) = adt_def.all_fields().enumerate().last().unwrap();
+                                let field_ty = field_def.ty(self.tcx, substs);
+                                if !matches!(field_ty.kind(), TyKind::Slice(..)) {
+                                    self.tcx.sess.span_fatal(
+                                        hir_param.ty_span,
+                                        "DST parameters are currently restricted to a reference to a struct whose last field is a slice.",
+                                    )
+                                }
+                                let len = emit
+                                    .array_length(len_t, None, var, member_idx as u32)
+                                    .unwrap();
+                                iter::once(var).chain(Some(len))
+                            }
+                            TyKind::Slice(..) | TyKind::Str => self.tcx.sess.span_fatal(
+                                hir_param.ty_span,
+                                "Straight slices are not yet supported, wrap the slice in a newtype.",
+                            ),
+                            // TODO: Is this needed?
+                            TyKind::Dynamic(..) => self.tcx.sess.span_fatal(
+                                hir_param.ty_span,
+                                "Trait objects are not supported.",
+                            ),
+                            _ => unreachable!(),
```

Comment:
I don't think this is unreachable. For example, here's a compiletest that hits it:

```rust
// build-pass
#![feature(extern_types)]
use spirv_std as _;
extern "C" {
    pub type Blah;
}
#[spirv(fragment)]
pub fn main(#[spirv(uniform)] x: &mut Blah) {}
```